torch/_functorch/aot_autograd.py

```diff
@@ -138,86 +138,365 @@ def posthook(grad_input, grad_output):
+    # This gives us info about every input, and what sort of mutation happened to it (if any)
+    input_info: List[InputAliasInfo]
+    # length = # user outputs + # intermediates that are returned as graph outputs
+    # This gives us info about every output (mostly around whether it aliases other tensors)
+    output_info: List[OutputAliasInfo]
+    # length = # outputs in the compiled forward. Equal to:
+    #   # mutated inps + # user outputs + # intermediate bases
+    #   (same as)
+    #   len(x for x in input_info if x.MutationType != MutationType.none) + len(output_info) + num_intermediate_bases
     # For every output *and* mutated input returned from the forward,
     # tells us whether or not the output should require gradients or not
-    requires_grad_out_info: List[bool]
-    # length: # fw outputs
-    aliased_output_info: List[OutputAliasInfo]
-
-def gen_alias_from_base(aliased_base_tensor, size, stride, storage_offset, target_meta_tensor):
-    # handle R2C and C2R
+    requires_grad_info: List[bool]
+    # length = len(x for x in input_info if x.MutationType != MutationType.none)
+    # This is pre-computed for perf: it contains the index of every element
+    # of input_info that corresponds to a mutation (data or metadata or both)
+    mutated_inp_indices: List[int]
+    # length = len(x for x in output_info if x.output_type != OutputType.non_alias)
+    # This is pre-computed for perf: it contains the index of every element
+    # of output_info that corresponds to an alias (either of an input or intermediate)
+    aliased_out_indices: List[int]
+    # length = the number of intermediate bases appended as outputs to the end of the forward graph.
+    # Note: this is not necessarily the same thing as:
+    #   len([x for x in output_info if x.output_type == OutputType.alias_of_intermediate])
+    # Because outputs might share a ._base, or an output's ._base might itself be
+    # another user output (in both cases, we won't redundantly append bases to the end of the graph)
+    num_intermediate_bases: int
+
+# This class exists because:
+# - the autograd.Function.forward() in aot autograd returns outputs that might alias inputs
+# - we only care about the metadata on those aliases, so we can regenerate them.
+#   We do not want them to participate in the autograd.Function.
+# We do that by wrapping them in an opaque class, so the autograd.Function
+# does not know to treat them as tensors.
+@dataclass(frozen=True)
+class TensorAlias:
+    alias: torch.Tensor
+    requires_grad: bool
+
+def has_same_metadata(t1, t2):
+    return t1.size() == t2.size() and t1.stride() == t2.stride() and t1.storage_offset() == t2.storage_offset()
+
+def is_data_mutation(m: MutationType):
+    # We should really turn on mypy
+    assert isinstance(m, MutationType)
+    return m in {MutationType.data, MutationType.data_and_metadata}
+
+def gen_alias_from_base(aliased_base_tensor, target_meta_tensor, target_requires_grad):
```

Comment:
I would love to know why you have to track target requires grad separately now

Reply:
bleh, tell me if you dislike this. I liked the idea of using an opaque `TensorAlias` to make it clear in the `autograd.Function.forward()` which outputs are aliases, and therefore shouldn't participate in autograd.

One downside of doing things that way, though, is that those outputs won't get `requires_grad=True` set after calling `outputs = CompiledFunction.apply(*inps)`.

The source of truth for which outputs require gradients was already tracked during tracing though, inside of `fw_metadata.requires_grad_info`. So that's where the requires_grad bool is pulled in from.